File: apis/ruby_api/stk_sequence.rb
require 'stk_env'

class Stk_sequence
	@@obj_map = {}
	def initialize(env,name,id,sequence_type,service_type,options,ref=nil)
		if ref == nil
			@_seq = Stksequence.stk_create_sequence(env.ref(),name,id,sequence_type,service_type,options);
		else
			@_seq = ref;
		end
		@_env = env
		@_id = Stksequence.stk_get_sequence_id(@_seq)
		@_seqptr = Stksequence.stk_seq_ptr_to_ulong_seq(@_seq)
		@@obj_map[@_seqptr] = self
	end
	def close()
		unmap();
		Stksequence.stk_destroy_sequence(@_seq)
	end
	def env()
		return @_env
	end
	def unmap()
		@@obj_map[@_seqptr] = nil
	end
	def hold()
		Stksequence.stk_hold_sequence(@_seq)
	end
	def id()
		return Stksequence.stk_get_sequence_id(@_seq);
	end
	def type()
		return Stksequence.stk_get_sequence_type(@_seq);
	end
	def count()
		return stk_number_of_sequence_elements(@_seq)
	end
	def release()
		stk_destroy_sequence(@_seq)
	end
	def ref()
		return @_seq
	end
	def iterate(clientd)
		n = Stksequence.stk_sequence_first_elem(@_seq)
		while Stksequence.stk_is_at_list_end(n) == 0 do
			dptr = Stksequence.stk_sequence_node_data(n)
			yield(self,dptr,Stksequence.stk_sequence_node_type(n),clientd)
			n = Stksequence.stk_sequence_next_elem(@_seq,n)
		end
	end
	def each(clientd)
		iterate(clientd)
	end
	def copy_array_to(data,user_type)
		v = Stksequence.new_voidDataArray(data.length)
		data.each_with_index do |c,i|
			Stksequence.voidDataArray_setitem(v,i,c)  # Set a value
		end
		Stksequence.stk_copy_chars_to_sequence(@_seq,v,data.length,user_type)
		Stksequence.delete_voidDataArray(v)
	end
	def copy_string_to(data,user_type)
		return Stksequence.stk_copy_string_to_sequence(@_seq,data,user_type)
	end
	def self.find(seqid)
		@@obj_map[seqid]
	end
end

File: apis/python_interface_helpers/stk_sequence.py
import stk_env
class stk_sequence:
	obj_map = {}
	def __init__(self,env,name,id,sequence_type,service_type,options,ref=None):
		if ref == None:
			self._seq = stk_create_sequence(env.ref(),name,id,sequence_type,service_type,options);
		else:
			self._seq = ref;
		self._env = env
		self._id = stk_get_sequence_id(self._seq)
		self._seqptr = stk_seq_ptr_to_ulong_seq(self._seq)
		self.__class__.obj_map[self._seqptr] = self
	def close(self):
		self.unmap();
		stk_destroy_sequence(self._seq)
	def env(self):
		return self._env
	def unmap(self):
		self.__class__.obj_map[self._seqptr] = None
	def hold(self):
		stk_hold_sequence(self._seq)
	def id(self):
		return stk_get_sequence_id(self._seq);
	def type(self):
		return stk_get_sequence_type(self._seq);
	def count(self):
		return stk_number_of_sequence_elements(self._seq)
	def release(self):
		stk_destroy_sequence(self._seq)
	def ref(self):
		return self._seq
	def iterate(self,cb,clientd):
		n = stk_sequence_first_elem(self._seq)
		while stk_is_at_list_end(n) == 0:
			dptr = stk_sequence_node_data(n)
			cb(self,dptr,stk_sequence_node_type(n),clientd)
			n = stk_sequence_next_elem(self._seq,n)
	def copy_array_to(self,data,user_type):
		v = new_voidDataArray(len(data))
		for i in range(0,len(data) - 1):
			voidDataArray_setitem(v,i,data[i])  # Set a value
		stk_copy_chars_to_sequence(self._seq,v,len(data),user_type)
		delete_voidDataArray(v)
	def copy_string_to(self,data,user_type):
		return stk_copy_string_to_sequence(self._seq,data,user_type)
	@classmethod
	def find(cls,seqid):
		if seqid in cls.obj_map:
			return cls.obj_map[seqid]
